src/engine/controls/cuecontrol.cpp

```diff
@@ -770,7 +888,98 @@ void CueControl::hotcueGotoAndPlay(HotcueControl* pControl, double value) {
     }
 }
 
-void CueControl::hotcueActivate(HotcueControl* pControl, double value) {
+void CueControl::hotcueGotoAndLoop(HotcueControl* pControl, double value) {
+    if (value == 0) {
+        return;
+    }
+
+    QMutexLocker lock(&m_mutex);
+    if (!m_pLoadedTrack) {
+        return;
+    }
+
+    CuePointer pCue(pControl->getCue());
+
+    // Need to unlock before emitting any signals to prevent deadlock.
+    lock.unlock();
+
+    if (!pCue) {
+        return;
+    }
+
+    double startPosition = pCue->getPosition();
+    if (startPosition == Cue::kNoPosition) {
+        return;
+    }
+
+    if (pCue->getType() == mixxx::CueType::Loop) {
+        seekAbs(startPosition);
+        setCurrentSavedLoopControlAndActivate(pControl);
+    } else if (pCue->getType() == mixxx::CueType::HotCue) {
+        seekAbs(startPosition);
+        setBeatLoop(startPosition, true);
+    } else {
+        return;
+    }
+
+    if (!isPlayingByPlayButton()) {
+        // cueGoto is processed asynchronously.
+        // avoid a wrong cue set if seek by cueGoto is still pending
+        m_bPreviewing = false;
+        m_iCurrentlyPreviewingHotcues = 0;
+        // don't move the cue point to the hot cue point in DENON mode
+        m_bypassCueSetByPlay = true;
+        m_pPlay->set(1.0);
+    }
+
+    m_pHotcueFocus->set(pControl->getHotcueNumber());
+}
+
+void CueControl::hotcueLoopToggle(HotcueControl* pControl, double value) {
+    if (value == 0) {
+        return;
+    }
+
+    if (!m_pLoadedTrack) {
+        return;
+    }
+
+    CuePointer pCue = pControl->getCue();
+
+    if (!pCue || pCue->getPosition() == Cue::kNoPosition) {
+        return;
+    }
+
+    switch (pCue->getType()) {
+    case mixxx::CueType::Loop: {
+        // The hotcue_X_loop_toggle CO was invoked for a saved loop, set it as
+        // active the first time this happens and toggle the loop_enabled state
+        // on subsequent invocations.
+        if (m_pCurrentSavedLoopControl != pControl) {
+            setCurrentSavedLoopControlAndActivate(pControl);
+        } else {
+            bool loopActive = pControl->getStatus() == HotcueControl::Status::Active;
+            setLoop(pCue->getPosition(), pCue->getEndPosition(), !loopActive);
+        }
+    } break;
+    case mixxx::CueType::HotCue: {
+        // The hotcue_X_loop_toggle CO was invoked for a hotcue. In that case,
+        // create a beatloop starting at the hotcue position. This is useful for
+        // mapping the CUE LOOP mode labeled on some controllers.
```

Comment:
```suggestion
        // Serato style controllers with a CUE LOOP mode to temporary use HotCues 1...8 for looping.
```

Reply:
Pioneer controllers also have this, e.g. the [DDJ-SB](https://img.audiofanzine.com/images/u/product/normal/pioneer-ddj-sb2-242421.png) (see the label for SHIFT+MANUAL LOOP above the pads).

Hotcues 1...8 is defined by the controller mapping. It works just as fine for other hotcue numbers.

Reply:
Right. The issue is that we need to make it more specific that the command is understandable without to much background. Isn't the cut at 8 mandatory for Serato cues? If yes, I would like to keep the comment, or an improved version.